timemachine/fe/single_topology.py

```diff
@@ -123,6 +122,78 @@ class DummyGroupAssignmentError(RuntimeError):
+def find_disabled_dummy_bonds_and_chiral_volumes(dummy_groups_ab, bond_graph_b, mol_b_chiral_idxs):
+    """
+    Find disabled dummy bonds and disabled chiral volumes given a dummy group assignment
+    """
+    induced_dummy_bonds = set()
+    dummy_atoms = set()
+    for anchor, dummy_group in dummy_groups_ab.items():
+        for dummy_atom_a in dummy_group:
+            dummy_atoms.add(dummy_atom_a)
+            # dummy-core bonds
+            if bond_graph_b.has_edge(anchor, dummy_atom_a):
+                induced_dummy_bonds.add(canonicalize_bond((anchor, dummy_atom_a)))
+            # dummy-dummy bonds
+            for dummy_atom_b in dummy_group:
+                if bond_graph_b.has_edge(dummy_atom_b, dummy_atom_a):
+                    induced_dummy_bonds.add(canonicalize_bond((dummy_atom_a, dummy_atom_b)))
+
+    full_dummy_bonds = set()
+    for u, v in bond_graph_b.edges():
+        if u in dummy_atoms or v in dummy_atoms:
+            full_dummy_bonds.add(canonicalize_bond((u, v)))
+
+    disabled_dummy_bonds = full_dummy_bonds.difference(induced_dummy_bonds)
+    disabled_dummy_chiral_volumes = set()
+    for c, i, j, k in mol_b_chiral_idxs:
+        if c in dummy_atoms or i in dummy_atoms or j in dummy_atoms or k in dummy_atoms:
+            if (
+                canonicalize_bond((c, i)) in disabled_dummy_bonds
+                or canonicalize_bond((c, j)) in disabled_dummy_bonds
+                or canonicalize_bond((c, k)) in disabled_dummy_bonds
+            ):
+                disabled_dummy_chiral_volumes.add((c, i, j, k))
+
+    return disabled_dummy_bonds, disabled_dummy_chiral_volumes
+
+
+def get_dummy_group_assignments_ordered_by_score(mol, core_atoms):
+    """
+    Generate a dummy group assigned based on score (lower is better). The heuristic tries to
+    minimize the # of core-bonds broken, followed by the number of chiral-volumes broken.
+
+    Parameters
+    ----------
+    mol: Chem.Mol
+        Input molecule
+
+    core_atoms: list of int
+        Atoms in the core of the molecule
+
+    Returns
+    -------
+    List of dummy_group_assignments, list of scores.
+        Sorted by ascending order based on list of scores.
+
+    """
+    bond_graph_b = convert_to_nx(mol)
+    mol_chiral_idxs = setup_all_chiral_atom_restr_idxs(mol, get_romol_conf(mol))
+    all_dummy_groups_ab = list(generate_dummy_group_assignments(bond_graph_b, core_atoms))
+
+    scores_for_mol = []
+    for dummy_groups_ab in all_dummy_groups_ab:
+        b_disabled_bonds, b_disabled_chiral_volumes = find_disabled_dummy_bonds_and_chiral_volumes(
+            dummy_groups_ab, bond_graph_b, mol_chiral_idxs
+        )
+        scores_for_mol.append((len(b_disabled_bonds), len(b_disabled_chiral_volumes)))
+
+    scores_for_mol = np.array(scores_for_mol, dtype=[("first", "int"), ("second", "int")])
+    perm = np.argsort(scores_for_mol, order=["first", "second"])
+
+    return [all_dummy_groups_ab[p] for p in perm], scores_for_mol[perm].tolist()
```

Comment:
nit: consider refactoring to avoid numpy array construction
```suggestion
    def compute_score(dummy_groups_ab):
        b_disabled_bonds, b_disabled_chiral_volumes = find_disabled_dummy_bonds_and_chiral_volumes(
            dummy_groups_ab, bond_graph_b, mol_chiral_idxs
        )
        return (len(b_disabled_bonds), len(b_disabled_chiral_volumes))

    sorted_pairs = sorted(
        [
            (compute_score(dummy_groups_ab), dummy_groups_ab)
            for dummy_groups_ab in generate_dummy_group_assignments(bond_graph_b, core_atoms)
        ]
    )

    return sorted_pairs
```
(If desirable to return separate lists, can replace `return sorted_pairs` with

```python
scores, sorted_dummy_groups = zip(*pairs)
return sorted_dummy_groups, scores
```